File: apps/web/src/features/chat/request.ts
import type {
  CatalogModel,
  CatalogTool,
  UploadedImageAttachment,
} from './types'

export const buildChatRequest = (
  sessionId: string,
  prompt: string,
  textModel: CatalogModel | null,
  imageTool: CatalogTool | null,
  attachments: UploadedImageAttachment[],
) => {
  return {
    session_id: sessionId,
    prompt,
    ...(attachments.length > 0 ? { attachments } : {}),
    ...(textModel
        ? {
            text_model: {
              model_config_id: textModel.model_config_id,
              model: textModel.model,
              provider: textModel.provider,
              runtime_provider: 'openai_compatible',
              source: textModel.source || undefined,
              display_name: textModel.display_name || undefined,
              type: textModel.type,
              input_modalities: textModel.input_modalities,
            },
          }
      : {}),
    ...(imageTool
      ? {
          tool_list: [
            {
              model_config_id: imageTool.model_config_id,
              id: imageTool.id,
              model: imageTool.model,
              provider: imageTool.provider,
              runtime_provider: 'openai_compatible',
              source: imageTool.source || undefined,
              display_name: imageTool.display_name || undefined,
              type: imageTool.type,
            },
          ],
        }
      : {}),
  }
}

File: crates/openchat-core/src/domain/request.rs
#[derive(Clone)]
pub struct SelectedTextModel {
    pub model_config_id: String,
    pub display_name: Option<String>,
    pub model_name: Option<String>,
    pub provider: Option<String>,
    pub runtime_provider: Option<String>,
    pub source: Option<String>,
    pub model_type: Option<String>,
    pub input_modalities: Vec<String>,
}

#[derive(Clone)]
pub struct UploadedAttachment {
    pub id: String,
    pub url: String,
    pub name: String,
    pub mime_type: String,
    pub size_bytes: usize,
    pub kind: Option<String>,
    pub extracted_text: Option<String>,
}

#[derive(Clone)]
pub struct SelectedTool {
    pub model_config_id: String,
    pub id: String,
    pub display_name: Option<String>,
    pub model_name: Option<String>,
    pub provider: Option<String>,
    pub runtime_provider: Option<String>,
    pub source: Option<String>,
    pub tool_type: String,
}

#[derive(Clone)]
pub struct ChatRequest {
    pub user_id: String,
    pub session_id: String,
    pub prompt: String,
    pub attachments: Vec<UploadedAttachment>,
    pub text_model: Option<SelectedTextModel>,
    pub tool_list: Vec<SelectedTool>,
}
